// pocket-sockets/build/src/TCPServer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TCPServer = void 0;
const net = __importStar(require("net"));
const tls = __importStar(require("tls"));
const Server_1 = require("./Server");
const TCPClient_1 = require("./TCPClient");
/**
 * TCP server implementation.
 */
class TCPServer extends Server_1.Server {
    constructor(serverOptions) {
        super(serverOptions);
        this.clientConnected = (socket) => {
            const client = new TCPClient_1.TCPClient({ bufferData: this.serverOptions.bufferData, port: this.serverOptions.port }, socket);
            this.addClient(client);
        };
        this.error = (error) => {
            this.serverError(Buffer.from(error.message));
        };
        this.serverCreate();
    }
    /**
     * Specifies how the server gets initialized, then creates the server with the specified options.
     */
    serverCreate() {
        var _a, _b, _c, _d, _e;
        const USE_TLS = this.serverOptions.cert != null;
        if (USE_TLS) {
            const tlsOptions = {
                cert: this.serverOptions.cert,
                key: this.serverOptions.key,
                requestCert: this.serverOptions.requestCert,
                rejectUnauthorized: this.serverOptions.rejectUnauthorized,
                ca: this.serverOptions.ca,
                handshakeTimeout: 30000,
            };
            this.server = tls.createServer(tlsOptions);
            (_a = this.server) === null || _a === void 0 ? void 0 : _a.on("secureConnection", this.clientConnected);
        }
        else {
            this.server = net.createServer();
            (_b = this.server) === null || _b === void 0 ? void 0 : _b.on("connection", this.clientConnected);
        }
        (_c = this.server) === null || _c === void 0 ? void 0 : _c.on("error", this.error);
        (_d = this.server) === null || _d === void 0 ? void 0 : _d.on("close", this.serverClosed);
        (_e = this.server) === null || _e === void 0 ? void 0 : _e.on("tlsClientError", (_err, socket) => socket.end());
    }
    /**
     * Starts a previously created server listening for connections.
     * Assumes the server is instantiated during object creation.
     */
    serverListen() {
        var _a;
        (_a = this.server) === null || _a === void 0 ? void 0 : _a.listen({
            host: this.serverOptions.host,
            port: this.serverOptions.port,
            ipv6Only: this.serverOptions.ipv6Only,
        });
    }
    serverClose() {
        if (this.server) {
            this.server.close();
        }
    }
}
exports.TCPServer = TCPServer;
